### ISSUE TITLE
[Suggestion] More Realistic Nuclear Science

### ISSUE THREAD

**Issue by Exaxxion** [GregTechCE/GregTech#1426]
Creating this ticket per @LAGIdiot's request here: https://github.com/GregTechCE/GregTech/pull/1414#issuecomment-766437332

## Overview
The existing recipes for processing Uraninite (represented as UO2 in GTCE), in which Aluminium or Magnesium are employed to strip the oxygen off of the Uranium atom, bear no resemblance to reality. The real process is quite involved, and I'll try to give as concise of an overview as I can here.

As a precursor, I should note that Pitchblende and Uraninite are different names for the same mineral, with the former name being antiquated. U238 and U235 don't appear as ores in nature. The way Pitchblende is processed in game, as well as its byproducts, are closer to reality than the other Uranium-containing ores. For the purposes here, I'll refer to the uranium oxide mineral as Uraninite.

## Reality in Depth
The following section is a lengthy description of the real processes used for conversion, enrichment, fabrication, and deconversion.

### Mining
The real process starts with crushing, leaching, drying, and filtering Uraninite into Yellowcake. This is mostly comprised of the geologically stable form Triuranium Octoxide `U3O8`, but also contains some Uranium Dioxide `UO2`, and Uranium Trioxide `UO3`. UO2 oxidizes with Oxygen to form U3O8, and at high temperatures UO3 decays into U3O8.

### Conversion
Conversion are processes by which uranium oxides are made into useful forms for enrichment or non-enriched uranium reactors.

Two main divergent processes exist from here: a "wet" and a "dry" process.

#### Dry Process
The dry process skips the chemical treatments and simply calcines the yellowcake to remove impurities before crushing it; you end up with U3O8.

#### Wet Process
The wet process is more involved: yellowcake is dissolved in nitric acid to create Uranyl Nitrate `UO2(NO3)2.6H2O`, which is purified via solvent extraction using tributyl phosphate dissolved in kerosene, and a washing (with dilute nitric acid) and evaporation process followed by calcination results in UO3. Alternately, treating Uranyl Nitrate with ammonia converts it into Ammonium Diuranate `(NH4)2U2O7` ("ADU"), which is then calcinated to produce UO3.

#### Process Convergence
The processes converge again here. The U3O8 or UO3, depending on the process, is reduced with hydrogen in a kiln:
`U3O8 + 2H2 -> 3UO2 + 2H2O`
`UO3 + H2 -> UO2 + H2O`

A second kiln is employed to convert the resulting Uranium Dioxide into Uranium Tetrafluoride `UF4` using gaseous Hydrogen Fluoride:
	`UO2 + 4HF -> UF4 + 2H2O`.
This form is good for Molten Salt Reactors, but it needs further processing for solid fuel reactors.

The next step is reacting the UF4 salt with fluorine gas to produce the intermediate form Uranium Pentafluoride `UF5`, then finally into gaseous (at high temperature) Uranium Hexafluoride `UF6`. Alternately, UF4 can be reduced with Magnesium to create metallic Uranium and the salt Magnesium Fluoride `MgF2`; given that it would have the natural ratios, this would result in >99% U238.

### Enrichment
UF6 is the crucial form for enriching Uranium. Centrifuging it in gaseous form takes advantage of the differing masses of the 235 and 238 isotopes of Uranium and the constant mass of Fluorine to separate the molecules formed with each isotope from each other. The ratio is quite small, as less than 1% of naturally occurring Uranium is isotopes other than 238.

The gas centrifuge process creates enriched and depleted UF6, which contain increased and reduced concentrations of U235, respectively. UF6 is enriched to concentrations of U235 high enough for its intended purpose, usually around 3-5% and not in excess of 20% for civilian reactors.

### Fabrication
For use in reactors, the enriched UF6 must be converted into UO2 powder and formed into pellets used to make fuel rods. This conversion also has differing "wet" and "dry" processes.

#### Dry Process
UF6 gas is mixed with steam in a two-stage vessel like a rotary kiln to create the powder Uranyl Fluoride:
	`UF6 + 2H2O -> UO2F2 + 4HF`.

Next, hydrogen diluted in steam is added to remove the remaining fluorine, ideally leaving pure UO2:
	`UO2F2 + H2 -> UO2 + 2HF`

However, this is not the whole picture. The uranium dioxide also reacts with hydrogen fluoride in the system, resulting in a mixture of reactants:
	`UO2 + 4HF -> UF4 + 2H2O`

Summing both equations results in the following:
	`2UO2F2 + 2H2 -> UO2 + UF4 + 2H2O`

#### Wet Process
UF6 is injected into water, creating a Uranyl Fluoride slurry. Ammonia is added to produce ADU, which is then filtered, dried, and heated to reduce it into UO2. Details are a bit fuzzy here but I believe this is the same calcination process used on ADU during conversion.

#### Fuel Rods
Once you have completed either prior process, you end up with enriched UO2 powder. This powder is conditioned with various additives to ensure the right properties for its target application. The resulting mixture is then pressed into pellets and sintered in a furnace at 1750C using a special atmosphere usually comprised of argon and hydrogen.

The resulting pellets are machined into precise shape and placed inside metal rods which are placed in a reactor as fuel.

### Storage and Deconversion

The depleted UF6 is mainly considered a waste product. It is usually kept in large secure containers, but it is a dangerous material to store since any failures of the container result in toxic and corrosive gases. It is useful to deconvert it back into U3O8, which also allows for reclaiming some of the HF used in the enrichment process.

For storage, one would vaporize the depleted UF6 in an autoclave with steam to produce uranyl fluoride:
	`UF6 + 2H2O -> UO2F2 + 4HF`
then add hydrogen and a lot of heat to finish the process and create the much safer U3O8 and reclaim even more hydrogen fluoride:
	`3UO2F2 + 2H2O + H2 -> U3O8 + 6HF`.

## Proposed Changes

Hope you enjoyed the chemistry lesson, we're both probably on a watchlist now.

The chemical processes noted above are quite involved and have many steps. In the interest in not massively overcomplicating things, we can drastically simplify the process. If we wanted to expand on it to include more things later, that would also be possible.

#### Add three new fluids (gases):

1. Uranium Hexafluoride
2. Enriched Uranium Hexafluoride (EUHF)
3. Depleted Uranium Hexafluoride (DUHF)

#### Replace the existing Uranium processing recipes with the following:

Machine | Recipe
------------|-----------
Chemical Reactor | Uraninite + Hydrofluoric Acid + Fluorine -> Uranium Hexafluoride + Water
Centrifuge | 10B UF6 -> 1B EUHF + 9B DUHF
Electrolyzer | 1B EUHF -> 1 U235 dust + 6B Fluorine
Electrolyzer | 1B DUHF -> 1 U238 dust + 6B Fluorine

#### Explanation
Skipping the notion of Yellowcake entirely, we start with Uraninite dust. We abstract away a ton of the processes to result in a chemical reactor recipe that produces Uranium Hexafluoride.

Centrifuging this approximates the refinement process and produces an unrealistically high - but reasonable for gameplay balance - mixture of enriched and depleted Uranium Hexafluoride.

Electrolysis is inaccurate, but would be a way the player would understand to separate the enriched and depleted UF6 into sensible quantities of U235 and U238.

**Comment by warjort** [comment]
You should checkout gregicality, it already kind of has what you describe including a gas centrifuge block to separate the isotopes.
It also only has pitchblende in its ore deposits with a fairly complicated process to get to the UF6.

**Comment by Exaxxion** [comment]
@warjort We had proposed altering this process in #1414; LAG was interested but wanted to defer that to a later time as a separate effort, since he wants to do other things like get radiation mechanics working.

Gregicality may have such features but that's unrelated to the development of GTCE itself.

**Comment by warjort** [comment]
Well, if you are going to reinvent the wheel, you should at least make sure it fits on the same axle as the one that already exists :-)

**Comment by Exaxxion** [comment]
Gregicality and GTCE have different philosophies and their desired implementations could be disjoint.

While it may be charitable to do so, there is no obligation on GTCE's part to ensure compatibility with existing features in addon mods. Gregicality devs could offer to PR the changes upstream to GTCE (if the implementation coincides with what GTCE wants), or override the implementation with their own.

**Comment by LAGIdiot** [comment]
@Exaxxion Thank you for creating separate issue for this. I really like depth of your explanation regarding this process, but I am not going to ask how you know all of it.

There is quite few things that we need to handle before we can implement this. We should get whole radiation and decay system going plus to have at least some abstract idea how to use this stuff (which means plans for nuclear reactor).

I would like to extend that even if we sometimes work together with our addons and we are grateful for them. We can't be shackled by them as GTCE is mod which is meant to be played not just API.

**Comment by DStrand1** [comment]
I have some ideas as to how to handle this:

**Radiation:**
- First off, we would want to add a Hazmat Suit that protects the player from radiation.
- We would want to look into how radioactive the different materials are, and implement a variety of effects that they could cause, depending on how bad they are.
- We could potentially have a mechanic similar to Mekanism v10 where if your reactor or storage (or whatever we choose to implement) has a leak, it could cause in-world effects and cause radiation.
- If we do the above, we could add a Geiger Counter tool to keep track of the radiation of items and the world

**Decay:**
I think this idea is more tentative than the above, but I still have some thoughts:
- Radioactive materials have a half-life, so we could add this as a mechanic, where your items will eventually decay into other items

I think this would be annoying to some players, so we should have this as a config option for sure. I still think it is a really cool idea, and I think it would be cool to add.

**Uses for radioactive materials:**
- We already have a Fusion Reactor multiblock (in addons, at least), we should add a Fission Reactor multiblock.
- I think it makes the most sense to have the Fission Reactor be before the Fusion Reactor, as thats more in-line with real world science involving radioactive materials.

There are two primary ways we can use this. Either:
1. Have it generate power on its own, similar to Nuclearcraft, but not very realistic.
2. Have it generate a LOT of steam from water, and power turbines, following the real world mechanism, as seen below:
![Nuclear reactor_0](https://user-images.githubusercontent.com/10861407/106378207-93a8e480-6368-11eb-83c9-f6e11b4575cd.png)

If we do this second option, which I think we should, we may want to look into the maximum rates of Steam Turbines, and possibly create a new higher-tier one to handle the massive amounts of Steam being used.

We would definitely need to add many more isotopes to use here, but I'm not sure how far exactly we would want to take that.

I would be more than happy to work on this issue, as it seems like a really fun and challenging problem to solve, and would really add a lot to GTCE.

**Comment by whatsapp2** [comment]
I think I'm getting a bit too carried away by the thought of having the fission reactor, but there are some random ideas that could be implemented:
-Actual nuclear waste that deposits in the nuclear reactor, Xenon-135 acts as a poison and stops the reactor
-Merging the Gregicality nuclear refinement process to this new fission reactor (processing of uraninite and pitchblende and reprocessing of depleted fuel rods)
-Some sort of irradiating samples with neutrons (like those irradiators from Nuclearcraft)
-Heat exchangers. This are very helpful to make some interactions possible like a fast breeder reactor that uses NaK alloy as a coolant, this hot NaK goes to heat exchangers and creates superheated steam that is then pumped into a separate heat exchanger to yield high pressure steam.
-High Pressure Turbines. Is a better way to convert different types of steam/plasma at higher Eu/t

**Comment by Deveritas** [comment]
I assume the uraninite processing would be following the chemistry overhaul, so the initial chemical reactor recipe would be either `U3O8 + 16HF + F2 -> 3UF6 + 8H2O`, allowing for a closed loop from the released fluoride and breaking down the water.

Re: DStrand1's comments, I think that if radiation is added, players should be given more ways to deal with the issue than just "wear this one specific suit of armor." I don't feel like that leads to meaningful gameplay - not in the same way that GT asks you to make other decisions about tradeoffs, as there really isn't a tradeoff when your choices are "wear this, move my entire base, or have really bad effects." (Side note, I've always wished to a more whimsical nuclear mod that has "nuclear sponge" enchants for sponges.)

But for power gen, I guess the real question is, what tier is this meant to take place in? Is this something that can help as far back as MV/HV/etc.? Mid-game? Or even as far forward as early fusion, when the better power-production recipes aren't available yet? I haven't played enough unmodified GTCE to know where the gaps are, unfortunately, so I can't answer that. 

Some ideas I had:
* A consequence for messing up your reactor, but a bit more realistic in that it's fail-safe and not fail-deadly: A multiblock reactor that has a Empty Corium Container (or some other name) that turns to another block when the multiblock fails, invalidating the structure.
* A Steam Processing Array multiblock. Maybe limited to only high-pressure steam machines. (Give those guys a use - does anyone really use them, instead of just skipping to LV?) Or at the very least, make sure there isn't a perverse incentive for "this is my nuclear reactor, and these are my 6,000 steam-age machines my reactor powers."
* Related to the previous idea, a new level of steam machines? I feel like this idea might be getting a bit too close to having parallel tiers - voltage and pressure - but maybe not. Side note, since the previous jump in steam machine tiers was from low pressure to high pressure, skipping medium pressure, I feel like these machines should be insane pressure, skipping extreme pressure. :)
